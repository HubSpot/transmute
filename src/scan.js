import curry from './curry';
import _scan from './internal/_scan';

/**
 * scan is similar to reduce, but returns a list of successive reduced values,
 * including the initial value
 *
 * @example
 * scan(0, (acc, val) => acc + val, [1, 2, 3]);
 * // returns List [ 0, 1, 3, 6 ]
 *
 * @example
 * scan(
 *   List(),
 *   (acc, val) => acc.push(val),
 *   Map({ one: 1, two: 2, three: 3 })
 * );
 * // returns List [ List [ ], List [ 1 ], List [ 1, 2 ], List [ 1, 2, 3 ] ]
 *
 * @param  {any} into
 * @param  {Function} operation
 * @param  {Iterable} subject   [description]
 * @return {Iterable}
 */
export default curry(_scan);
