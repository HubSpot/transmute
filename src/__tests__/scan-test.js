import { List, Map, Seq, Set } from 'immutable';
import scan from '../scan';

describe('transmute/scan', () => {
  it('scans an Array', () => {
    const scanner = jest.fn((acc, n) => acc.push(n + 1));
    expect(scan(List(), scanner, [1, 2, 3])).toMatchSnapshot();
    expect(scanner.mock.calls.length).toBe(3);
    expect(scanner.mock.calls[0]).toEqual([List(), 1, 0, [1, 2, 3]]);
    expect(scanner.mock.calls[1]).toEqual([List.of(2), 2, 1, [1, 2, 3]]);
    expect(scanner.mock.calls[2]).toEqual([List.of(2, 3), 3, 2, [1, 2, 3]]);
  });

  it('scans a List', () => {
    expect(
      scan(List(), (acc, n) => acc.push(n + 1), List.of(1, 2, 3))
    ).toMatchSnapshot();
  });

  it('scans a Map', () => {
    expect(
      scan(
        Map(),
        (acc, n, k) => acc.set(k, n + 1),
        Map({ one: 1, two: 2, three: 3 })
      )
    ).toMatchSnapshot();
  });

  it('scans an Object', () => {
    expect(
      scan(List(), (acc, n) => acc.push(n + 1), { one: 1, two: 2, three: 3 })
    ).toMatchSnapshot();
  });

  it('scans a Seq', () => {
    expect(
      scan(Seq(), (acc, n) => acc.concat([n + 1]), Seq([1, 2, 3]))
    ).toMatchSnapshot();
  });

  it('scans a Set', () => {
    expect(
      scan(Set(), (acc, n) => acc.add(n + 1), Set.of(1, 2, 3))
    ).toMatchSnapshot();
  });

  it('scans to a primitive', () => {
    expect(scan(0)((acc, n) => acc + n)(List.of(1, 2, 3))).toMatchSnapshot();
  });
});
