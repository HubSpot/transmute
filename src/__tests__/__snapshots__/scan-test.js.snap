// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmute/scan scans a List 1`] = `
Immutable.List [
  Immutable.List [],
  Immutable.List [
    2,
  ],
  Immutable.List [
    2,
    3,
  ],
  Immutable.List [
    2,
    3,
    4,
  ],
]
`;

exports[`transmute/scan scans a Map 1`] = `
Immutable.List [
  Immutable.Map {},
  Immutable.Map {
    "one": 2,
  },
  Immutable.Map {
    "one": 2,
    "two": 3,
  },
  Immutable.Map {
    "one": 2,
    "two": 3,
    "three": 4,
  },
]
`;

exports[`transmute/scan scans a Seq 1`] = `
Immutable.List [
  Immutable.Seq [],
  Immutable.Seq [
    2,
  ],
  Immutable.Seq […],
  Immutable.Seq […],
]
`;

exports[`transmute/scan scans a Set 1`] = `
Immutable.List [
  Immutable.Set [],
  Immutable.Set [
    2,
  ],
  Immutable.Set [
    2,
    3,
  ],
  Immutable.Set [
    2,
    3,
    4,
  ],
]
`;

exports[`transmute/scan scans an Array 1`] = `
Immutable.List [
  Immutable.List [],
  Immutable.List [
    2,
  ],
  Immutable.List [
    2,
    3,
  ],
  Immutable.List [
    2,
    3,
    4,
  ],
]
`;

exports[`transmute/scan scans an Object 1`] = `
Immutable.List [
  Immutable.List [],
  Immutable.List [
    2,
  ],
  Immutable.List [
    2,
    3,
  ],
  Immutable.List [
    2,
    3,
    4,
  ],
]
`;

exports[`transmute/scan scans to a primitive 1`] = `
Immutable.List [
  0,
  1,
  3,
  6,
]
`;
